fish-rust/src/wutil/encoding.rs

```diff
@@ -0,0 +1,37 @@
+extern "C" {
+    pub fn wcrtomb(s: *mut libc::c_char, wc: libc::wchar_t, ps: *mut mbstate_t) -> usize;
+    pub fn mbrtowc(
+        pwc: *mut libc::wchar_t,
+        s: *const libc::c_char,
+        n: usize,
+        p: *mut mbstate_t,
+    ) -> usize;
+}
+
+extern "C" {
+    #[cfg(any(feature = "apple_libc", feature = "bsd_libc"))]
+    fn ___mb_cur_max() -> usize;
+    #[cfg(feature = "glibc")]
+    fn __ctype_get_mb_cur_max() -> usize;
+}
+
+#[cfg(any(feature = "apple_libc", feature = "bsd_libc"))]
+#[allow(non_snake_case)]
+pub fn MB_CUR_MAX() -> usize {
+    unsafe { ___mb_cur_max() }
+}
+#[cfg(feature = "glibc")]
+#[allow(non_snake_case)]
+pub fn MB_CUR_MAX() -> usize {
+    unsafe { __ctype_get_mb_cur_max() }
+}
+
+// HACK This should be mbstate_t from libc but that's not exposed.  Since it's only written by
+// libc, we define it as opaque type that should be large enough for all implementations.
+pub type mbstate_t = [u64; 16];
+pub fn zero_mbstate() -> mbstate_t {
+    [0; 16]
+}
+
+// HACK This should be the MB_LEN_MAX macro from libc but that's not easy to get.
+pub const AT_LEAST_MB_LEN_MAX: usize = 32;
```

Comment:
I think the value here is safe to use, but I've been planning on adding build-time constant retrieval to `build.rs`. This is a good candidate.

Reply:
how should we retrieve the constants, I think the most robust way requires us to run a small C program.

Reply:
32 seems fine to me, it's very conservative. The largest value I can find in the wild is 16.